### DISCUSSION TITLE
Multithreading and modifying variables in callbacks

### DISCUSSION THREAD

**Discussion by williamleong** [bitcraze/discussions#1128] [Q&A] [answered]
Callbacks are added with `p2pRegisterCB` and `cpxRegisterAppMessageHandler`. I am wondering if they are multi-threaded and modified variables need to be protected against race conditions, as I don't see any mutexes or semophores being used in the examples.

**Comment by ataffanel** [comment, marked as answer]
These callback functions are called from the context of the communication tasks, so yes, you need to protect communication between these callback and the rest of your code if you are running in another context.

These protection might, in some case, not be required like when one side is only reading and another side only writing a memory that can be written in one instructions, that might be what happens in some of our example. There might also be some oversight in others... If you find an example that has a clear data-race bug (like a read-modify-write on a shared variable), please feel free to make a ticket in the repos that hosts the example.

**Reply by williamleong** [reply]
Thanks @ataffanel that seems reasonable and thanks for clarifying.
